File: agent-adapter/python/turboflow_adapter/registry.py
"""Backend Registry — Factory + Registry pattern."""

from __future__ import annotations

from typing import Callable

from turboflow_adapter.backend import AgentBackend
from turboflow_adapter.errors import UnknownBackendError
from turboflow_adapter.logger import get_logger
from turboflow_adapter.types import BackendInfo

log = get_logger("tf-adapter.registry")

BackendFactory = Callable[[], AgentBackend]


class BackendRegistry:
    """Central registry for all backends. Supports dynamic registration."""

    def __init__(self) -> None:
        self._factories: dict[str, BackendFactory] = {}
        self._instances: dict[str, AgentBackend] = {}
        self._register_builtins()

    def _register_builtins(self) -> None:
        from turboflow_adapter.backends.claude import ClaudeBackend
        from turboflow_adapter.backends.aider import AiderBackend
        from turboflow_adapter.backends.openhands import OpenHandsBackend
        from turboflow_adapter.backends.strands import StrandsBackend

        self.register("claude", ClaudeBackend)
        self.register("aider", AiderBackend)
        self.register("openhands", OpenHandsBackend)
        self.register("strands", StrandsBackend)

    def register(self, id: str, factory: BackendFactory) -> None:
        self._factories[id] = factory
        log.debug("Backend registered: %s", id)

    def get(self, id: str) -> AgentBackend:
        if id in self._instances:
            return self._instances[id]

        factory = self._factories.get(id)
        if not factory:
            raise UnknownBackendError(id, self.list_ids())

        instance = factory()
        self._instances[id] = instance
        return instance

    def has(self, id: str) -> bool:
        return id in self._factories

    def list_ids(self) -> list[str]:
        return list(self._factories.keys())

    def list_all(self) -> list[BackendInfo]:
        return [self.get(id).to_info() for id in self.list_ids()]


# Singleton
registry = BackendRegistry()


def create_backend(id: str) -> AgentBackend:
    """Factory function — get a backend by ID."""
    return registry.get(id)

File: agent-adapter/ts/src/registry.ts
// =============================================================================
// TurboFlow Agent Adapter — Backend Registry (Factory + Registry Pattern)
//
// Central registry for all backends. Supports dynamic registration so
// third-party backends can be added without modifying core code.
// =============================================================================

import type { BackendId, BackendInfo } from './types.js';
import type { AgentBackend } from './backend.js';
import { UnknownBackendError } from './errors.js';
import { Logger } from './logger.js';

import { ClaudeBackend } from './backends/claude.js';
import { AiderBackend } from './backends/aider.js';
import { OpenHandsBackend } from './backends/openhands.js';

const log = new Logger('registry');

type BackendFactory = () => AgentBackend;

class BackendRegistry {
  private factories = new Map<string, BackendFactory>();
  private instances = new Map<string, AgentBackend>();

  constructor() {
    // Register built-in backends
    this.register('claude', () => new ClaudeBackend());
    this.register('aider', () => new AiderBackend());
    this.register('openhands', () => new OpenHandsBackend());
  }

  /** Register a new backend factory */
  register(id: string, factory: BackendFactory): void {
    this.factories.set(id, factory);
    log.debug(`Backend registered: ${id}`);
  }

  /** Get or create a backend instance */
  get(id: string): AgentBackend {
    // Return cached instance
    const cached = this.instances.get(id);
    if (cached) return cached;

    // Create new instance
    const factory = this.factories.get(id);
    if (!factory) {
      throw new UnknownBackendError(id, this.listIds());
    }

    const instance = factory();
    this.instances.set(id, instance);
    return instance;
  }

  /** Check if a backend is registered */
  has(id: string): boolean {
    return this.factories.has(id);
  }

  /** List all registered backend IDs */
  listIds(): string[] {
    return [...this.factories.keys()];
  }

  /** List all registered backends with metadata */
  listAll(): BackendInfo[] {
    return this.listIds().map((id) => this.get(id).toInfo());
  }

  /** Clear cached instances (useful for testing) */
  clearInstances(): void {
    this.instances.clear();
  }
}

/** Singleton registry */
export const registry = new BackendRegistry();

/**
 * Factory function — get a backend by ID.
 * Shorthand for registry.get().
 */
export function createBackend(id: BackendId | string): AgentBackend {
  return registry.get(id);
}
